design-proposals/vsock/vsock.md

```diff
@@ -0,0 +1,92 @@
+*   To simplify the migration story, the CID will be unique cluster wide. Therefore the CID will be retained after migration. (With 100 000 k8s nodes we would still have enough CIDs to run more than 20 000 VMs per node). The CID has a type of `uint32`, which is large enough to satisfy a cluster-wide unique allocation)
+
+![vsock flow](vsock.png)
+
+#### API changes
+
+A `autoattachVsock` flag will be added to the VMI spec at `spec.domain.devices` which defaults to `false` to stay backward compatible. Also a `VsockCID` field will be added to the VMI status as `status.vsockCID` for CID allocation and lookup.
+
+### Phase 2: websocket access via the apiserver
+
+This follows exactly the same architecture as vnc/console. `GET /virtualmachineinstances/&lt;name>/vsock?port=123` will be added to the REST-API and it will return a websocket connection to the guest, if something is listening on the guest. Then, for example, node-local or control-plane services can establish a gRPC connection on top of the websocket connection to the guest service and invoke RPC calls.
+
+By default none of our aggregated cluster-roles will get RBAC access to this endpoint.
+
+Creating a websocket connection via this REST-API means that all traffic goes
+through the kubernetes apiserver. Therefore the same scale considerations apply
+like for `exec`, `port-forward`, `vnc` and so forth.
+
+### Security
+
+Note: The `dev/vhost-vsock` device is **NOT NEEDED** to connect or bind to a `vsock` port. This means that it is slightly more challenging to secure it.
+
+The following measure will therefore be put in place:
+
+*   The whole `vsock` features will live behind a feature gate
+*   By default the first 1024 ports of a vsock device are privileged. Services trying  to connect to those require `CAP_NET_BIND_SERVICE`.
+*   We aim to block the `AF_VSOCK` `socket` syscall in containerd (https://github.com/containerd/containerd/issues/7442), however if that will happen is not sure. It is therefore right now the responsibility of the vendor to apply the required seccomp change to all unprivileged ports, or to verify that the `CAP_NET_BIND_SERVICE` privilege is handed out carefully enough
+*   RBAC access to vsock devices will not be added to any of the default roles.
+*   virt-launcher will not auto-assign CIDs to ensure that virt-handler has an easy way of tracking the identity without races. While this still allows virt-launcher to fake-use an assigned CID, it eliminates possible assignment races which attackers could make use-of to redirect vsock calls.
+
+### Possible next steps
+
+*   Allow running `qemu-guest-agent` over virtio-vsock too
+*   Introduce a general purpose mechanism via gRPC in virt-handler which allows efficient and scalable collection of version and readiness of registered third-party agents without having to alter virt-handler, or having to run additional node-local services. Also move the `qemu-guest-agent` polling behind this mechanism.
```

Comment:
Nit `qemu-quest-agent` use case is mentioned above.

Reply:
Could you explain what you mean? Running `qemu-guest-agent` on vsock is an explicit non-goal, but I wanted to outline how a possible future integration in a more general guest-agent mechanism could look like.

Reply:
Hm, they are two different points though:
 * This proposal does not change code-paths in virt-handler to allow running the qemu-guest-agent via vsock
 * Another point would be introducing a common polling mechanism, where no extra handling for the qemu-guest-agent would be there, it would use a common extensible mechanism.
 
 That are just some rough ideas to highlight that while qemu-guest-agent is not tackled  in this proposal, it is not in the way of doing so later.

Reply:
Now I got you. I still feel like the last part of the second point is redundant. Your comment explained it very well, maybe we can incorporate it into the points?